[SPARK-10652][SPARK-10742][Streaming] Set meaningful job descriptions for all streaming jobs

tdas opened the pull request:
Here is the screenshot after adding the job descriptions to threads that run receivers and the scheduler thread running the batch jobs.
## All jobs page
- Added job descriptions with links to relevant batch details page
  ![image](https://cloud.githubusercontent.com/assets/663212/9924165/cda4a372-5cb1-11e5-91ca-d43a32c699e9.png)
## All stages page
- Added stage descriptions with links to relevant batch details page
  ![image](https://cloud.githubusercontent.com/assets/663212/9923814/2cce266a-5cae-11e5-8a3f-dad84d06c50e.png)
## Streaming batch details page
- Added the +details link
  ![image](https://cloud.githubusercontent.com/assets/663212/9921977/24014a32-5c98-11e5-958e-457b6c38065b.png)

tdas committed:
Set job descriptions for all streaming jobs

tdas commented:
@zsxwing Can you take a look at the changes.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

zsxwing commented on streaming/src/main/scala/org/apache/spark/streaming/ui/BatchPage.scala:
```diff
@@ -207,16 +207,23 @@ private[ui] class BatchPage(parent: StreamingTab) extends WebUIPage("batch") {
+          <div class="stage-details collapsed">
+            <pre>{stageInfo.details}</pre>
+          </div>
+        }
```
`else` is missing.

zsxwing commented on streaming/src/main/scala/org/apache/spark/streaming/ui/BatchPage.scala:
```diff
@@ -207,16 +207,23 @@ private[ui] class BatchPage(parent: StreamingTab) extends WebUIPage("batch") {
-
-    val lastStageName = lastStageInfo.map(_.name).getOrElse("(Unknown Stage Name)")
-    val lastStageDescription = lastStageData.flatMap(_.description).getOrElse("")
+    Text(lastStageInfo.map { _.name }.getOrElse("(Unknown Stage)"))
```
This line is not used.

SparkQA commented:
[Test build #42562 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42562/consoleFull) for   PR 8791 at commit [`e8e7894`](https://github.com/apache/spark/commit/e8e7894ab6e7fa9e1ca4ac8b4d692d966cbef361).

SparkQA commented:
[Test build #42562 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42562/console) for   PR 8791 at commit [`e8e7894`](https://github.com/apache/spark/commit/e8e7894ab6e7fa9e1ca4ac8b4d692d966cbef361).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42562/
Test FAILed.

tdas committed:
Added link to streaming UI in desc of streaming jobs

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

tdas committed:
Addressed comments

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #42573 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42573/consoleFull) for   PR 8791 at commit [`7206f2e`](https://github.com/apache/spark/commit/7206f2e9d4c42edf5a6a231b5cbde7309a62f041).

SparkQA commented:
[Test build #42573 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42573/console) for   PR 8791 at commit [`7206f2e`](https://github.com/apache/spark/commit/7206f2e9d4c42edf5a6a231b5cbde7309a62f041).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42573/
Test FAILed.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42572/
Test FAILed.

tdas committed:
More fixes

AmplabJenkins commented:
Merged build triggered.

tdas commented:
@zsxwing I addressed comments and made another update. Now the job description supports having links in them, and that is used to link to the batch details page.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #42574 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42574/consoleFull) for   PR 8791 at commit [`b15012f`](https://github.com/apache/spark/commit/b15012f3a2b4f4205723d6ea37bf22bfe7c300f7).

SparkQA commented:
[Test build #42574 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42574/console) for   PR 8791 at commit [`b15012f`](https://github.com/apache/spark/commit/b15012f3a2b4f4205723d6ea37bf22bfe7c300f7).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42574/
Test FAILed.

tdas committed:
Remove println

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #42592 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42592/consoleFull) for   PR 8791 at commit [`fd0d435`](https://github.com/apache/spark/commit/fd0d435ecce8ca06e78615339a1e9ab713d422d3).

SparkQA commented:
[Test build #42592 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42592/console) for   PR 8791 at commit [`fd0d435`](https://github.com/apache/spark/commit/fd0d435ecce8ca06e78615339a1e9ab713d422d3).
 * This patch **fails MiMa tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42592/
Test FAILed.

tdas committed:
Use scala.xml's own escaping

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #42599 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42599/consoleFull) for   PR 8791 at commit [`b5cb02e`](https://github.com/apache/spark/commit/b5cb02e3b2193b16b4b10117bdeb92933248dfb2).

SparkQA commented:
[Test build #42599 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42599/console) for   PR 8791 at commit [`b5cb02e`](https://github.com/apache/spark/commit/b5cb02e3b2193b16b4b10117bdeb92933248dfb2).
 * This patch **fails MiMa tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42599/
Test FAILed.

zsxwing commented on streaming/src/main/scala/org/apache/spark/streaming/StreamingContext.scala:
```diff
@@ -198,7 +198,7 @@ class StreamingContext private[streaming] (
   private var state: StreamingContextState = INITIALIZED
 
-  private val startSite = new AtomicReference[CallSite](null)
+  private[streaming] val startSite = new AtomicReference[CallSite](null)
```
Looks it's better to add a new method like
```
private[streaming] def getStartSite: String = startSite.get()
``` 
rather than exposing `AtomicReference`. In addition, this one won't break MiMa tests.

zsxwing commented:
LGTM except my comment about `startSite`.

tdas committed:
Addressed comments, and added base URL prepending to links in description

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #42659 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42659/consoleFull) for   PR 8791 at commit [`d2ab91d`](https://github.com/apache/spark/commit/d2ab91d2e79c61d7fe137f7cd89bc291bfb8944c).

SparkQA commented:
[Test build #42659 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42659/console) for   PR 8791 at commit [`d2ab91d`](https://github.com/apache/spark/commit/d2ab91d2e79c61d7fe137f7cd89bc291bfb8944c).
 * This patch **fails MiMa tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42659/
Test FAILed.

tdas commented:
This is blocked on #8781 . Once that is merged in both master and 1.5, I will finish this PR by adding unit tests.

tdas committed:
Merge remote-tracking branch 'apache-github/master' into SPARK-10652

tdas committed:
Updated test

tdas commented:
@andrewor14

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #42797 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42797/consoleFull) for   PR 8791 at commit [`21c90ea`](https://github.com/apache/spark/commit/21c90eafa20cb8433675401372b4c5c223cf0957).

SparkQA commented:
[Test build #42797 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42797/console) for   PR 8791 at commit [`21c90ea`](https://github.com/apache/spark/commit/21c90eafa20cb8433675401372b4c5c223cf0957).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42797/
Test FAILed.

tdas committed:
Added check for anchors, and unit tests

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #42815 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42815/consoleFull) for   PR 8791 at commit [`4a45c82`](https://github.com/apache/spark/commit/4a45c825705d6af72c0bec3673be231cab852410).

SparkQA commented:
[Test build #42815 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42815/console) for   PR 8791 at commit [`4a45c82`](https://github.com/apache/spark/commit/4a45c825705d6af72c0bec3673be231cab852410).
 * This patch **fails RAT tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42815/
Test FAILed.

tdas committed:
Add license

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #42825 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42825/consoleFull) for   PR 8791 at commit [`4a0aba8`](https://github.com/apache/spark/commit/4a0aba8d995659ef9a0f92081e18278f29dd4259).

SparkQA commented:
[Test build #42825 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42825/console) for   PR 8791 at commit [`4a0aba8`](https://github.com/apache/spark/commit/4a0aba8d995659ef9a0f92081e18278f29dd4259).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42825/
Test FAILed.

tdas committed:
style fix

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #42826 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42826/consoleFull) for   PR 8791 at commit [`f9524c6`](https://github.com/apache/spark/commit/f9524c61fd56ba07099e96db86f95cff88520aa4).

SparkQA commented:
[Test build #42826 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42826/console) for   PR 8791 at commit [`f9524c6`](https://github.com/apache/spark/commit/f9524c61fd56ba07099e96db86f95cff88520aa4).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42826/
Test FAILed.

zsxwing commented:
retest this please

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #42840 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42840/consoleFull) for   PR 8791 at commit [`f9524c6`](https://github.com/apache/spark/commit/f9524c61fd56ba07099e96db86f95cff88520aa4).

zsxwing commented on core/src/main/scala/org/apache/spark/ui/UIUtils.scala:
```diff
@@ -395,4 +397,55 @@ private[spark] object UIUtils extends Logging {
+        xml \\ "a" flatMap { _.attributes } filter { _.key == "href" } map { _.value.toString }
+      if (allLinks.exists { ! _.startsWith ("/") }) {
+        throw new IllegalArgumentException(
+          "Links in job descriptions must be relative:\n" + allLinks.mkString("\n\t"))
```
I think both `/link` and `link` are called relative links. But `/link` is called `root-relative link`. Maybe we should use `Links in job descriptions must be root-relative:\n` here?

SparkQA commented:
[Test build #42840 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42840/console) for   PR 8791 at commit [`f9524c6`](https://github.com/apache/spark/commit/f9524c61fd56ba07099e96db86f95cff88520aa4).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42840/
Test FAILed.

tdas committed:
fixed test

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #42849 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42849/consoleFull) for   PR 8791 at commit [`2fbfd1f`](https://github.com/apache/spark/commit/2fbfd1f215f80072b15b65d6e43beb82156acf5d).

SparkQA commented:
[Test build #42849 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42849/console) for   PR 8791 at commit [`2fbfd1f`](https://github.com/apache/spark/commit/2fbfd1f215f80072b15b65d6e43beb82156acf5d).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42849/
Test PASSed.

tdas commented:
@andrewor14 Could you take a look at this.

andrewor14 commented on streaming/src/main/scala/org/apache/spark/streaming/scheduler/JobScheduler.scala:
```diff
@@ -190,10 +191,20 @@ class JobScheduler(val ssc: StreamingContext) extends Logging {
+        val batchLinkText = s"[output operation ${job.outputOpId}, batch time ${formattedTime}]"
+
+        ssc.sc.setJobDescription(
+          s"""Streaming job from <a href="$batchUrl">$batchLinkText</a>""")
```
this is assuming the UI is the only consumer of this? I suppose that's OK since no one uses `JobLogger` anymore

tdas replied:
Whoever uses it will get the data as text. Only the UI would actually process it as a link. Nothing breaks.

srowen replied:
It looks like this caused a cosmetic issue in the tooltips though: https://github.com/apache/spark/pull/11845

andrewor14 commented on core/src/test/scala/org/apache/spark/ui/UIUtilsSuite.scala:
```diff
@@ -0,0 +1,65 @@
+    )
+  }
+
+  def verify(desc: String, expected: Elem, errorMsg: String = "", baseUrl: String = ""): Unit = {
```
private

andrewor14 commented on core/src/main/scala/org/apache/spark/ui/UIUtils.scala:
```diff
@@ -395,4 +397,56 @@ private[spark] object UIUtils extends Logging {
+   * Returns HTML rendering of a job or stage description. It will try to parse the string as HTML
+   * and make sure that it only contains anchors with relative links. Otherwise, the whole string
+   * will rendered as a simple escaped text.
+   */
```
can you add a paragraph on the security implications here? (What happens if I put in a `<script>` tag or something, or if I reference a malicious link)

andrewor14 replied:
actually, what happens if I put a script tag?

andrewor14 commented:
core parts LGTM

tdas committed:
Addressed Andrew's comments

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

zsxwing commented:
LGTM

SparkQA commented:
[Test build #42878 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42878/consoleFull) for   PR 8791 at commit [`606b284`](https://github.com/apache/spark/commit/606b2847f49c5a4e769af88a0054199050d2f880).

tdas commented:
Thanks @andrewor14 @zsxwing, will merge this to master and 1.5 when tests pass.

SparkQA commented:
[Test build #42878 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42878/console) for   PR 8791 at commit [`606b284`](https://github.com/apache/spark/commit/606b2847f49c5a4e769af88a0054199050d2f880).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42878/
Test PASSed.
